google/resource_dns_record_set.go

```diff
@@ -55,6 +57,147 @@ func resourceDnsRecordSet() *schema.Resource {
 	}
 }
 
+// This catches most invalid DNS names, though it might not catch some invlaid non-ASCII names
+// it does not implement the TOASCII functionality defined in https://tools.ietf.org/html/rfc3454
+// See:
+// https://tools.ietf.org/html/rfc1034 DOMAIN NAMES - CONCEPTS AND FACILITIES
+// https://tools.ietf.org/html/rfc3490 Internationalizing Domain Names in Applications (IDNA)
+// https://tools.ietf.org/html/rfc3454 Preparation of Internationalized Strings ('stringprep')
+func ValidateDNSRRsetName(i interface{}, k string) (s []string, errors []error) {
+	name := i.(string)
+	errors = append(errors, checkName(name)...)
+	return
+}
+
+func checkName(name string) (errors []error) {
+
+	if len(name) > 255 {
+		errors = append(errors, fmt.Errorf("DNS name length %d greater than 254",
+			len(name)))
+	}
+
+	// https://tools.ietf.org/html/rfc3490#section-3.1 Requirements
+	// Whenever dots are used as label separators, the following
+	// characters MUST be recognized as dots: U+002E (full stop),
+	// U+3002 (ideographic full stop), U+FF0E (fullwidth full
+	// stop), U+FF61 (halfwidth ideographic full stop).
+	re, err := regexp.Compile("[\u002e\u3002\uFF0E\uFF61]$")
```

Comment:
Totally up to you, but here's a way you could make this a bit more concise if you wanted:
```
if re, err := regexp.Compile("[\u002e\u3002\uFF0E\uFF61]$"); err != nil {
  errors = append(errors, fmt.Errorf("Internal error compiling regexp: %s", err))
} else if !re.MatchString(name) {
  errors = append(errors, fmt.Errorf("DNS name must be fully qualified with a trailing dot"))
}

Reply:
This is by design. It probably does not matter in most cases, but I'd like to keep the more expensive regular expression compilation outside the loop. I'll add a comment section that emphasizes this.